src/plugins/dataflow/components/dataflow-tile.tsx

```diff
@@ -157,17 +163,24 @@ export default class DataflowToolComponent extends BaseComponent<IProps, IDatafl
 
   private pairNodesToAttributes = () => {
     const model = this.getContent();
+    // dataSet looks like
+    // Time_Quantized | Time_Actual | Node 1 | Node 2 | Node 3 etc
+    //    0           |  0          | val    | val    |  val
+    model.addNewAttrFromNode(0, "Time_Quantized");
+    model.addNewAttrFromNode(1, "Time_Actual");
 
-    //#1 check nodes on tile against dataset attributes, if already there do nothing, otherwise write.
     model.program.nodes.forEach((n) => {
       model.addNewAttrFromNode(n.id, n.name);
     });
 
-    //#2 check dataset attributes against nodes on tile, if an attribute is not on the tile - remove it.
+    // dataset attributes against nodes on tile, if an attribute is not on the tile - remove it.
     const dataSet = model.dataSet;
     const dataSetAttributes = dataSet.attributes;
+
     dataSetAttributes.forEach((attribute, idx) => {
-      model.removeAttributesInDatasetMissingInTile(attribute.id);
+      if (idx > 1) { //skip 0 and 1 index because those attribute are Time
```

Comment:
Yet another place where the assumption of two time attributes is coded. Perhaps the `getAttributeIdForNode()` function I suggested above can be somewhere accessible to this code as well.

Reply:
So I tried to change `getAttributeIdForNode()` from a private function into an exportable function. I'd have to pass the `dataSet` as a model, was having a hard time getting it to compile and read off the attributes since it had to be outside the scope of the Class `DataflowProgram` .  Also was getting many long typecheck errors when passing the `dataSet` calling it from `dataflow-tile.tsx`

Tried pairing with Joe and couldn't get it to work. Would be curious to know how we could get it to work if you have time to pair. 

But I left it as is, because I'd also argue that while  making this change, you could remove two lines related to the index logic (182 and 184), you'd still add a line for obtaining the attribute by calling `getAttributeIdForNode().`

Reply:
The most logical "somewhere accessible to this code as well", i.e. somewhere accessible to the `DataflowProgram` and the `DataflowTile`, is the `DataflowContent`. The `getAttributeIdForNode()` method could easily be a method on the content. Looking more closely, however, `getAttributeIdForNode()` doesn't work for the `DataflowTile`'s use case because it requires the reverse mapping (i.e. subtracting rather than adding two). The simplest approach at this point might be to put a constant like `export const kTimeAttributeCount = 2;`  in a common location like `DataflowContent` and use that in both places, so at least the value can't be changed in one place without changing it in the other. I leave it to your discretion how to resolve things from here.

Reply:
Sure that's a good idea, I'll add a constant offset count that both files are referring to, push up and merge.